include/irr/asset/IImageAssetHandlerBase.h

```diff
@@ -31,10 +37,207 @@ class IImageAssetHandlerBase : public virtual core::IReferenceCounted
+
+				return singleRowImage;
+			};
+
+			core::smart_refctd_ptr<ICPUImage> inputSingleRowImage = createImage(inputFormat);
+			core::smart_refctd_ptr<ICPUImage> outputSingleRowImage = createImage(outputFormat, false);
+
+			auto attachedRegion = outputSingleRowImage->getRegions().begin();
+
+			state.inImage = inputSingleRowImage.get();
+			state.outImage = outputSingleRowImage.get();
+			state.inOffset = { 0, 0, 0 };
+			state.inBaseLayer = 0;
+			state.outOffset = { 0, 0, 0 };
+			state.outBaseLayer = 0;
+			state.extent = attachedRegion->getExtent();
+			state.layerCount = attachedRegion->imageSubresource.layerCount;
+			state.inMipLevel = attachedRegion->imageSubresource.mipLevel;
+			state.outMipLevel = attachedRegion->imageSubresource.mipLevel;
+
+			if (!convertFilter.execute(&state))
+				os::Printer::log("Something went wrong while converting the row!", ELL_WARNING);
+
+			return outputSingleRowImage;
+		}
+
+		/*
+			Patch for not supported by OpenGL R8_SRGB formats.
+			Input image needs to have all the regions filled 
+			and texel buffer attached as well.
+		*/
+
+		static inline core::smart_refctd_ptr<ICPUImage> convertR8ToR8G8B8Image(core::smart_refctd_ptr<ICPUImage> image)
+		{
+			constexpr auto inputFormat = EF_R8_SRGB;
+			constexpr auto outputFormat = EF_R8G8B8_SRGB;
+
+			using CONVERSION_FILTER = CConvertFormatImageFilter<inputFormat, outputFormat>;
+
+			core::smart_refctd_ptr<ICPUImage> newConvertedImage;
+			{
+				auto referenceImageParams = image->getCreationParameters();
+				auto referenceBuffer = image->getBuffer();
+				auto referenceRegions = image->getRegions();
+				auto referenceRegion = referenceRegions.begin();
+				const auto newTexelOrBlockByteSize = asset::getTexelOrBlockBytesize(outputFormat);
+
+				auto newImageParams = referenceImageParams;
+				auto newCpuBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(referenceBuffer->getSize() * newTexelOrBlockByteSize);
+				auto newRegions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(referenceRegions.size());
+
+				for (auto newRegion = newRegions->begin(); newRegion != newRegions->end(); ++newRegion)
+				{
+					*newRegion = *(referenceRegion++);
+					newRegion->bufferOffset = newRegion->bufferOffset * newTexelOrBlockByteSize;
+				}
+
+				newImageParams.format = outputFormat;
+				newConvertedImage = ICPUImage::create(std::move(newImageParams));
+				newConvertedImage->setBufferAndRegions(std::move(newCpuBuffer), newRegions);
+
+				CONVERSION_FILTER convertFilter;
+				CONVERSION_FILTER::state_type state;
+
+				state.inImage = image.get();
+				state.outImage = newConvertedImage.get();
+				state.inOffset = { 0, 0, 0 };
+				state.inBaseLayer = 0;
+				state.outOffset = { 0, 0, 0 };
+				state.outBaseLayer = 0;
+
+				for (auto newAttachedRegion = newConvertedImage->getRegions().begin(); newAttachedRegion != newConvertedImage->getRegions().end(); ++newAttachedRegion)
```

Comment:
no `convertFilter` will already take care of this without any problems, you can just set the state to cover the whole image above

Reply:
you told me once I need to fly through regions to perform it per mipmaps

Reply:
when I asked you what about mipmaped layered images

Reply:
there is no actually field such as "mip level count" in state, how can I do it?

Reply:
you're iterating over alll regions, ever

you should be iterating over mipmaps
```cpp
for (uint32_t mip=0u; params.mipLevelCount; i++)
{
//fill out state
execute(&state);
}
```

the filter itself will get the appropriate regions and do operations on them (and clip the areas properly by itself as well)